Respect network scope in prometheus plugin

philipnrmn opened the pull request:
This PR fixes an issue where metrics were pulled via the local agent IP address instead of the task's IP, which could prevent it working in container mode. 

Instead of joining the port up to the local IP, we now use the ip_addresses field in container status to collect the container's declared ip address, which is present in all networking modes. 

Unit tests: 7eb7ee45bab4af3449d2a072bafec9fd6424b93f
Implementation: ce332c60545af689133f323768028f25c319ad6b

The final five commits update the existing tests to use a sample IP address instead of their original value.

philipnrmn committed:
Add missing info to portlabel README

philipnrmn committed:
Add a unit test for three network modes

This commit adds a test case which tests prometheus-labeled ports in all three
network modes: host, bridge and container.

Add ports and labels to test case

philipnrmn committed:
Implement proper IP address retrieval

philipnrmn committed:
Fix tasklabelViaIndex

philipnrmn committed:
Fix tasklabelAlternatePath

philipnrmn committed:
Fix portlabel

philipnrmn committed:
Fix tasklabelViaName

philipnrmn committed:
Fix tasklabelIndexPriority

philipnrmn commented:
The failing test on this PR is a flake.

branden reviewed (dismissed):
✨

gracedo commented on plugins/inputs/prometheus/prometheus.go:
```diff
@@ -202,7 +202,7 @@ func (p *Prometheus) GetAllURLs() (map[string]URLAndAddress, error) {
 		}
 
-		for _, url := range getMesosTaskPrometheusURLs(tasks, p.mesosHostname) {
+		for _, url := range getMesosTaskPrometheusURLs(tasks) {
```
Would it make sense to call `getTaskIP` above this line and pass in the IP, keeping this and the other function signatures the same? It would also only run the `getTaskIP` logic once as opposed to in multiple places.

I also did a cursory search of `p.mesosHostname` and it doesn't look like it's used anywhere anymore so maybe we can do a clean up and remove that field and [`getMesosHostname`](https://github.com/dcos/telegraf/pull/58/files#diff-7126ee4cf2b69635e3c56ae1cdb36e20R384)

philipnrmn replied:
The change this PR makes is that instead of using the hostname to form the metrics source URL, we now use the task's ip address. So we no longer need to pass the hostname in. 

Each task can potentially have a different IP address. Since this method retrieves URLs from multiple tasks, passing it a single IP address won't work - it needs to use the IP address from each task. 

WRT your second point, you're quite right - I had thought it was used in `getMesosClient`, but it isn't. I'll remove it.

gracedo replied:
Ahh, I see, missed that loop the first time round :). It looks like both `getEndpointsFromTaskPorts` and `getEndpointFromTaskLabels` get passed the same task, so maybe `getTaskIP` can just be called once on the task and passed into both of those functions to clean that up a bit and avoid the extra call - but it's not a big deal and I won't block on it 😸

philipnrmn committed:
Remove unneeded mesosHostname field and getMesosHostname func

philipnrmn commented:
The DC/OS integration tests have caught an issue! In bridge mode, we need either to use `host_ip:host_port` or `container_ip:container_port`. Right now we're using `container_ip:host_port`, which doesn't work.

philipnrmn committed:
Revert "Remove unneeded mesosHostname field and getMesosHostname func"

This reverts commit 1a80b57044e59d0bc823598e0054fbcc15bfcdd8.

philipnrmn committed:
Correct assumptions in test cases

philipnrmn committed:
Add method to retrieve metrics hostname for task

The correct hostname for a task is determined by the network-scope label on the
port. If set to 'container', the hostname is the first IP address in the last
container status. If set to 'host' or absent, the hostname is the same as the
local node.

philipnrmn committed:
Call metrics hostname method in relevant methods

gracedo commented:
retest this please

philipnrmn commented:
the `blkio_cfq` input in `dcos_containers` test is a flake and unrelated to code changed in this PR. I'm going to merge this now.

vespian commented:
I wonder if networks isolation (e.g. some iptables rules) in calico will not interfere here - i.e. it will not be possible to make a connection from host to the container that is attached to the SDN network.

philipnrmn commented:
@vespian thanks for pointing this out. We can't readily work around it, so we'll document this as a limitation: https://jira.mesosphere.com/browse/DCOS-56350
